chain/chain/src/chain.rs

```diff
@@ -1488,27 +1498,135 @@ impl Chain {
         Ok(())
     }
 
-    /// Process a received or produced block, and unroll any orphans that may depend on it.
+    /// Start process a received or produced block. This function will process block asynchronously.
+    /// It preprocesses the block by verifying that the block is valid and ready to process, then
+    /// schedules the work of applying chunks in rayon thread pool. The function will return before
+    /// the block processing is finished.
+    /// This function is used in conjunction with the function postprocess_ready_blocks, which checks
+    /// if any of the blocks in processing has finished applying chunks to finish postprocessing
+    /// these blocks that are ready.
     /// `block_processing_artifacts`: Callers can pass an empty object or an existing BlockProcessingArtifact.
     ///              This function will add the effect from processing this block to there.
-    pub fn process_block(
+    /// `apply_chunks_done_callback`: This callback will be called after apply_chunks are finished
+    ///              (so it also happens asynchronously in the rayon thread pool). Callers can
+    ///              use this callback as a way to receive notifications when apply chunks are done
+    ///              so it can call postprocess_ready_blocks.
+    pub fn start_process_block_async(
         &mut self,
         me: &Option<AccountId>,
         block: MaybeValidated<Block>,
         provenance: Provenance,
         block_processing_artifacts: &mut BlockProcessingArtifact,
-    ) -> Result<Option<Tip>, Error> {
+        apply_chunks_done_callback: ApplyChunkCallback,
```

Comment:
Regarding the `apply_chunks_done_callback`, I need it because apply_chunks need to send a message to client actor but it doesn't know client actor's address. I was deciding whether this callback should be a field in Chain and in Client and hide it from the interface of functions of processing blocks, or should be passed to these functions. Eventually I decided that this should be exposed to the caller in the interface, to make it clear to the caller if a function can potentially start the processing of some blocks but not finishing it. So the caller knows that it must remember to call postprocess_ready_blocks later. This doesn't matter in production code, because ClientActor will call postprocess_ready_blocks on every actix message. However, this is very useful in the Client testing code, when the test must always remember to call postprocess_ready_blocks whenever there might be blocks in processing.